File: mbox_project_remind/ex26_8/ex26_8_6.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define BUFFER_SIZE 1024
#define SEARCH_PREFIX "From: "
#define PREFIX_LEN 6

void	print_help(char *prog_name)
{
	printf("=== How to Use ===\n");
	printf("[%s] [mbox File]\n", prog_name);
	printf("\nThis program will extract sender emails from an mbox file\n");
	printf("command '-h' of '--help' to show this message\n");
}

int	ext_sender_and_copy(char *from_line, char **email)
{
	char	*start = NULL;
	char	*end = NULL;

	if ((start = strchr(from_line, '<')) != NULL) {
		start++;
		if ((end = strchr(from_line, '>')) == NULL)
			return 1;
	} else if ((start = strchr(from_line, ' ')) != NULL) {
		start++;
		if ((end = strchr(from_line, '\n')) == NULL)
			return 1;
	} else
		return 1;

	int	interval = end - start;
	*email = malloc(interval + 1);
	if (*email == NULL)
		return 1;
	strncpy(*email, start, interval);
	(*email)[interval] = '\0';
	return 0;
}

void	count_line(FILE *fp)
{
	// 到達していない場合、fpのポインタをそこまで進める
	// 動かすのはfget関数が返すファイル構造体のポインタなので
		// fpを動かすわけでは無い。よってasteriskは1つ

	char	c;
	// 条件式の演算子に注意
	while ((c = fgetc(fp)) != EOF && c != '\n')
		;
}

int	main(int argc, char **argv)
{
	if (argc != 2) {
		fprintf(stderr, "Argument Error\n");
		return 1;
	}

	if (strcmp(argv[1], "-h") == 0 || strcmp(argv[1], "--help") == 0) {
		print_help(argv[0]);
		return 0;
	}

	const char	*file_name = argv[1];
	FILE		*fp = fopen(file_name, "r");
	if (fp == NULL) {
		fprintf(stderr, "Cannot Open File\n");
		return 1;
	}

	time_t	start_time = clock();

	char	buffer[BUFFER_SIZE];
	char	*email = NULL;
	int	line_num = 0;
	while (fgets(buffer, sizeof(buffer), fp) != NULL) {
		size_t	len = strlen(buffer);
		// fpを動かす前に検証する必要がある
		// bufferの最後はNULL文字だから、そのまま行くと必ず動いてしまう
		if (len > 0 && buffer[len - 1] != '\n')
			count_line(fp);
		line_num++;

		if (strncmp(buffer, SEARCH_PREFIX, PREFIX_LEN) == 0) {
			if (ext_sender_and_copy(buffer, &email) == 1) {
				fprintf(stderr, "Cannot extract email\n");
				free(email);
				fclose(fp);
				return 1;
			}
			printf("%d: %s\n", line_num, email);
			free(email);
		}
	}

	time_t	end_time = clock();
	double	cpu_time = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;

	// 標準出力への表示を制限している
	fprintf(stderr, "=== Statistics ===\n");
	fprintf(stderr, "\nTotal Lines: %d\n", line_num);
	fprintf(stderr, "Prosessing Time: %.3f s\n", cpu_time);


	fclose(fp);
	return 0;
}

// clock()の実装

File: mbox_project_remind/ex26_8/ex26_8_6.py
import sys
import re
import time

"""
今回の目的は.cとcpc時間を比較する事
その為デコードは行わず、純粋にFrom: からemailを取り出す処理を行う

12-10:
    1. このコードのレビュー
    2. テスト実行によるCPU時間比較のためのShellScript作成
"""


def print_help(prog_name):
    print("=== How to Use ===", file=sys.stderr)
    print(f"[{prog_name}] [.mbox file]", file=sys.stderr)
    print("", file=sys.stderr)
    print("This program extract sender emails from an mbox file", file=sys.stderr)
    print("Command '-h' or '--help' to show this message", file=sys.stderr)

def print_status(line_num, cpu_time):
    print("=== Statistics ===", file=sys.stderr)
    print("", file=sys.stderr)
    print(f"Total Lines: {line_num}", file=sys.stderr)
    print(f"Processing Time: {cpu_time:.3f} s", file=sys.stderr)

def ext_sender(from_line):
    if from_line:
        # []の中では、'.'は特殊文字ではないので、\は不要
        email = re.search(r"[\w.-]+@[\w.-]+", from_line)
        return email.group() if email else None

def main():
    if len(sys.argv) != 2:
        print("Argument Error", file=sys.stderr)
        return 1
    if "-h" == sys.argv[1] or "--help" == sys.argv[1]:
        print_help(sys.argv[0])
        return 0

    file_name = sys.argv[1]
    line_num = 0
    # pythonでcpu時間を出すときの関数
    start_clock = time.process_time()

    try:
        with open(file_name, "r") as file:
            for line in file:
                line_num += 1
                # startswithは正規表現を使わない
                # 先頭文字列を確認するだけ
                if line.startswith("From: "):
                    email = ext_sender(line)
                    print(f"{line_num}: {email}") if email else print(f"{line_num}: Extract Failed")
    except FileNotFoundError:
        print(f"Error: Cannot Open {file_name}", file=sys.stderr)
        return 1
    except Exception as e:
        print(f"Unexpected Error Detected: {e}", file=sys.stderr)
        return 1

    end_clock = time.process_time()
    cpu_time = end_clock - start_clock
    print_status(line_num, cpu_time)
    return 0


"""
このソースコードが直接実行された場合、mainの戻り値をプログラムが返す
.cのmain関数的な実装をする際の記述
もっとも、.pyでは必ずしも必要ではないので、明確な意図があることが多い。詳しくはNotionを参照
"""

if __name__ == '__main__':
    sys.exit(main())
